File: lib/data/database/daos/achievement_dao.dart
import 'package:drift/drift.dart';
import '../app_database.dart';
import '../tables.dart';

part 'achievement_dao.g.dart';

@DriftAccessor(tables: [Achievements])
class AchievementDao extends DatabaseAccessor<AppDatabase> with _$AchievementDaoMixin {
  AchievementDao(super.db);

  Stream<List<AchievementEntity>> watchAll() =>
      (select(achievements)..orderBy([(t) => OrderingTerm(expression: t.unlockedAt, mode: OrderingMode.desc)]))
          .watch();

  Future<List<AchievementEntity>> getAll() =>
      (select(achievements)).get();

  Future<bool> isUnlocked(String achievementId) async {
    final result = await (select(achievements)..where((t) => t.achievementId.equals(achievementId))).get();
    return result.isNotEmpty;
  }

  Future<void> unlock(String achievementId) async {
    final existing = await (select(achievements)..where((t) => t.achievementId.equals(achievementId))).get();
    if (existing.isEmpty) {
      await into(achievements).insert(
        AchievementsCompanion.insert(
          id: Value('ach_${DateTime.now().microsecondsSinceEpoch}'),
          achievementId: achievementId,
        ),
      );
    }
  }
}

File: app/src/main/java/com/salesquest/sales_quest/data/dao/AchievementDao.kt
package com.salesquest.sales_quest.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.salesquest.sales_quest.data.entity.AchievementEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface AchievementDao {

    @Query("SELECT * FROM achievements ORDER BY unlockedAt DESC")
    fun watchAll(): Flow<List<AchievementEntity>>

    @Query("SELECT * FROM achievements")
    suspend fun getAll(): List<AchievementEntity>

    @Query("SELECT COUNT(*) FROM achievements WHERE achievementId = :achievementId")
    suspend fun isUnlocked(achievementId: String): Int

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun unlock(achievement: AchievementEntity)

    @Query("DELETE FROM achievements")
    suspend fun clearAll()
}
